plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -647,32 +654,58 @@ public Iterator<TableColumnsMetadata> streamTableColumns(ConnectorSession sessio
         else {
             schemaTableNames = ImmutableList.of(prefix.toSchemaTableName());
         }
-        return schemaTableNames.stream()
-                .flatMap(tableName -> {
+
+        BlockingQueue<TableColumnsMetadata> columnQueue = new LinkedBlockingQueue<>();
+        AtomicInteger completedTasks = new AtomicInteger(0);
+
+        schemaTableNames.parallelStream().forEach(tableName ->
+                metadataListingExecutor.submit(() -> {
                     try {
                         if (redirectTable(session, tableName).isPresent()) {
-                            return Stream.of(TableColumnsMetadata.forRedirectedTable(tableName));
+                            columnQueue.add(TableColumnsMetadata.forRedirectedTable(tableName));
                         }
 
                         Table icebergTable = catalog.loadTable(session, tableName);
                         List<ColumnMetadata> columns = getColumnMetadatas(icebergTable.schema());
-                        return Stream.of(TableColumnsMetadata.forTable(tableName, columns));
+                        columnQueue.add(TableColumnsMetadata.forTable(tableName, columns));
                     }
                     catch (TableNotFoundException e) {
                         // Table disappeared during listing operation
-                        return Stream.empty();
                     }
                     catch (UnknownTableTypeException e) {
                         // Skip unsupported table type in case that the table redirects are not enabled
-                        return Stream.empty();
                     }
                     catch (RuntimeException e) {
                         // Table can be being removed and this may cause all sorts of exceptions. Log, because we're catching broadly.
                         log.warn(e, "Failed to access metadata of table %s during streaming table columns for %s", tableName, prefix);
-                        return Stream.empty();
                     }
-                })
-                .iterator();
+
+                    // Tasks must reach this point as all Exceptions are caught above
+                    completedTasks.getAndIncrement();
+                }));
+
+        return new Iterator<>() {
```

Comment:
when implementing Iterator, lean on `com.google.common.collect.AbstractIterator`. you then do the logic once